Firestore/Example/Tests/Remote/FSTRemoteEventTests.mm

```diff
@@ -43,85 +44,159 @@ @interface FSTRemoteEventTests : XCTestCase
+                changes:(NSArray<FSTWatchChange *> *)watchChanges {
   FSTWatchChangeAggregator *aggregator =
-      [[FSTWatchChangeAggregator alloc] initWithSnapshotVersion:testutil::Version(3)
-                                                  listenTargets:listens
-                                         pendingTargetResponses:outstanding];
-  [aggregator addWatchChanges:watchChanges];
+      [[FSTWatchChangeAggregator alloc] initWithTargetMetadataProvider:_targetMetadataProvider];
+
+  NSMutableArray<FSTBoxedTargetID *> *targetIDs = [NSMutableArray array];
+  [targetMap enumerateKeysAndObjectsUsingBlock:^(FSTBoxedTargetID *targetID,
+                                                 FSTQueryData *queryData, BOOL *stop) {
+    [targetIDs addObject:targetID];
+    [_targetMetadataProvider setSyncedKeys:existingKeys forQueryData:queryData];
+  }];
+
+  [outstandingResponses
+      enumerateKeysAndObjectsUsingBlock:^(FSTBoxedTargetID *targetID, NSNumber *count, BOOL *stop) {
+        for (int i = 0; i < count.intValue; ++i) {
+          [aggregator recordTargetRequest:targetID];
+        }
+      }];
+
+  for (FSTWatchChange *change in watchChanges) {
+    if ([change isKindOfClass:[FSTDocumentWatchChange class]]) {
+      [aggregator handleDocumentChange:(FSTDocumentWatchChange *)change];
+    } else if ([change isKindOfClass:[FSTWatchTargetChange class]]) {
+      [aggregator handleTargetChange:(FSTWatchTargetChange *)change];
+    }
+  }
+
+  [aggregator handleTargetChange:[[FSTWatchTargetChange alloc]
+                                     initWithState:FSTWatchTargetChangeStateNoChange
+                                         targetIDs:targetIDs
+                                       resumeToken:_resumeToken1
+                                             cause:nil]];
+
   return aggregator;
 }
 
+- (FSTRemoteEvent *)
+remoteEventAtSnapshotVersion:(FSTTestSnapshotVersion)snapshotVersion
+                   targetMap:(NSDictionary<FSTBoxedTargetID *, FSTQueryData *> *)targetMap
+        outstandingResponses:
+            (nullable NSDictionary<FSTBoxedTargetID *, NSNumber *> *)outstandingResponses
+                existingKeys:(DocumentKeySet)existingKeys
+                     changes:(NSArray<FSTWatchChange *> *)watchChanges {
+  FSTWatchChangeAggregator *aggregator = [self aggregatorWithTargetMap:targetMap
+                                                  outstandingResponses:outstandingResponses
+                                                          existingKeys:existingKeys
+                                                               changes:watchChanges];
+  return [aggregator remoteEventAtSnapshotVersion:testutil::Version(snapshotVersion)];
+}
+
 - (void)testWillAccumulateDocumentAddedAndRemovedEvents {
-  FSTDocument *doc1 = FSTTestDoc("docs/1", 1, @{ @"value" : @1 }, NO);
-  FSTDocument *doc2 = FSTTestDoc("docs/2", 2, @{ @"value" : @2 }, NO);
+  NSDictionary<FSTBoxedTargetID *, FSTQueryData *> *targetMap =
```

Comment:
There's now some pretty subtle interaction in here and it's not entirely clear to me how these pieces fit together.

Could you add some comments about how they're supposed to work? For example, on top of `listenForTargets:`, what does the resulting map represent? Listens previously established?

Another piece that's missing: what if you omit a target id from `listensForTargets:` and then create an `FSTDocumentWatchChange` referring to that target id. I think that would simulate an update to a previously listened-to target, but I'm not entirely sure today and I'll certainly have no idea in 6 months :-).

Reply:
I renamed the `listen` and `limboListen` helpers to `queryDataForTarget`, which should make it make a bit more obvious what is going on.

If you use a target in the tests that is not included in the `targetMap`, then `FSTTestTargetMetadataProvider` will hit an assert telling you that the target is unknown.

I have added a bunch of method comments to the top of this test file and commented the first test case as an example. I hope that clears up some of the confusion.